// .ignored_x402-hono/dist/cjs/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  paymentMiddleware: () => paymentMiddleware
});
module.exports = __toCommonJS(src_exports);
var import_schemes = require("x402/schemes");
var import_shared = require("x402/shared");
var import_verify = require("x402/verify");
function paymentMiddleware(payToAddress, routes, facilitator) {
  const { verify, settle } = (0, import_verify.useFacilitator)(facilitator);
  const x402Version = 1;
  const routePatterns = (0, import_shared.computeRoutePatterns)(routes);
  return async function paymentMiddleware2(c, next) {
    const matchingRoute = (0, import_shared.findMatchingRoute)(routePatterns, c.req.path, c.req.method.toUpperCase());
    if (!matchingRoute) {
      return next();
    }
    const { price, network } = matchingRoute.config;
    const { description, mimeType, maxTimeoutSeconds, outputSchema, customPaywallHtml, resource } = matchingRoute.config.config || {};
    const atomicAmountForAsset = (0, import_shared.processPriceToAtomicAmount)(price, network);
    if ("error" in atomicAmountForAsset) {
      throw new Error(atomicAmountForAsset.error);
    }
    const { maxAmountRequired, asset } = atomicAmountForAsset;
    const resourceUrl = resource || c.req.url;
    const paymentRequirements = [
      {
        scheme: "exact",
        network,
        maxAmountRequired,
        resource: resourceUrl,
        description: description ?? "",
        mimeType: mimeType ?? "application/json",
        payTo: payToAddress,
        maxTimeoutSeconds: maxTimeoutSeconds ?? 300,
        asset: (asset == null ? void 0 : asset.address) ?? "",
        outputSchema,
        extra: asset == null ? void 0 : asset.eip712
      }
    ];
    const payment = c.req.header("X-PAYMENT");
    const userAgent = c.req.header("User-Agent") || "";
    const acceptHeader = c.req.header("Accept") || "";
    const isWebBrowser = acceptHeader.includes("text/html") && userAgent.includes("Mozilla");
    if (!payment) {
      if (isWebBrowser) {
        const displayAmount = typeof price === "string" || typeof price === "number" ? Number(price) : Number(price.amount) / 10 ** price.asset.decimals;
        const html = customPaywallHtml ?? (0, import_shared.getPaywallHtml)({
          amount: displayAmount,
          paymentRequirements: (0, import_shared.toJsonSafe)(paymentRequirements),
          currentUrl: c.req.path,
          testnet: network === "base-sepolia"
        });
        return c.html(html, 402);
      }
      return c.json(
        {
          error: "X-PAYMENT header is required",
          accepts: paymentRequirements,
          x402Version
        },
        402
      );
    }
    let decodedPayment;
    try {
      decodedPayment = import_schemes.exact.evm.decodePayment(payment);
    } catch (error) {
      return c.json(
        {
          error: error instanceof Error ? error : new Error("Invalid or malformed payment header"),
          accepts: paymentRequirements,
          x402Version
        },
        402
      );
    }
    const verification = await verify(decodedPayment, paymentRequirements[0]);
    if (!verification.isValid) {
      return c.json(
        {
          error: new Error(verification.invalidReason),
          accepts: paymentRequirements,
          payer: verification.payer,
          x402Version
        },
        402
      );
    }
    await next();
    try {
      const settlement = await settle(decodedPayment, paymentRequirements[0]);
      if (settlement.success) {
        c.header(
          "X-PAYMENT-RESPONSE",
          JSON.stringify({
            success: true,
            transaction: settlement.transaction,
            network: settlement.network
          })
        );
      }
    } catch (error) {
      return c.json(
        {
          error: error instanceof Error ? error : new Error("Failed to settle payment"),
          accepts: paymentRequirements,
          x402Version
        },
        402
      );
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  paymentMiddleware
});
